### PR TITLE
feat(server): background self-update on startup

### PR DESCRIPTION
**Pull request by WonderMr**
## Summary

Adds an opt-out background self-updater so the MCP server keeps its install
current without blocking startup. On launch a daemon thread fast-forwards the
repo to the target branch and rebuilds the vector stores in a subprocess; the
pulled code takes effect on the next start (for per-session stdio servers, the
next spawn). The heavy reindex runs in the background of the current session so
the next one starts fast. No os.execv, no pip install.

## Behavior and safety

- Enabled by default; disable with `AGENTS_AUTO_UPDATE=0`.
- Acts only when the checked-out branch is `AGENTS_AUTO_UPDATE_BRANCH` (default
  `main`) and the working tree is clean, fast-forward only. It is a no-op on any
  other branch, so local development is never disturbed.
- Never switches branches, merges, or rebases.
- A failed reindex (for example, broken new code) is rolled back with
  `git reset --hard` to the pre-merge commit.
- Cross-process non-blocking lock so multiple concurrent stdio servers do not
  race on git; an optional throttle avoids repeated network fetches.
- Offline or any error is logged and the server keeps serving the current code.

## Config (env vars)

| Variable | Default | Purpose |
| :-- | :-- | :-- |
| `AGENTS_AUTO_UPDATE` | `1` | Master on/off switch |
| `AGENTS_AUTO_UPDATE_REMOTE` | `origin` | Git remote to track |
| `AGENTS_AUTO_UPDATE_BRANCH` | `main` | Only updates when this branch is checked out |
| `AGENTS_AUTO_UPDATE_TIMEOUT` | `30` | Seconds per git op |
| `AGENTS_AUTO_UPDATE_INTERVAL` | `900` | Throttle network checks (0 = every start) |
| `AGENTS_AUTO_UPDATE_REINDEX_TIMEOUT` | `600` | Seconds for the reindex subprocess |

## Files

- New: `src/self_update.py`, `src/reindex.py`, `tests/test_self_update.py`
- Changed: `src/engine/config.py`, `src/server.py`, `env.example`, `README.md`

## Testing

- 15 new tests in `tests/test_self_update.py` over temporary git repos (reindex
  mocked, no model load): up-to-date, fast-forward applies, wrong-branch skip,
  dirty skip, diverged skip, fetch failure fail-open, reindex failure rollback,
  disabled, lock contention, throttle.
- Full fast suite: 469 passed. The 6 failures in `test_language.py` are
  pre-existing (langdetect not installed in the local venv) and unrelated.
- `python -m src.reindex` runs to completion (exit 0).
- On `feat/auto_update` the updater is a verified no-op (SKIPPED_WRONG_BRANCH),
  working tree untouched.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Background auto-update daemon that non-blockingly fast-forwards the repo and rebuilds vector stores at startup, with safe rollback on failed reindex and continued serving on errors.
  * CLI entry to run a manual rebuild of vector stores.

* **Configuration**
  * New env vars to enable/control auto-update, choose remote/branch, and tune timeouts and throttle intervals.

* **Documentation**
  * README updated with auto-update usage and configuration.

* **Tests**
  * Added tests covering updates, rollback, throttling, and edge cases.

### CODE REVIEW

File: src/self_update.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+            logger.info("Auto-update: working tree has uncommitted changes; skipping.")
+            return UpdateStatus.SKIPPED_DIRTY
+
+        head = _run_git(["rev-parse", "HEAD"], repo_root, git_timeout)
+        old_sha = head.stdout.strip() if head.returncode == 0 else ""
+
+        # Step 4 — fetch the target branch.
+        try:
+            fetch = _run_git(["fetch", remote, branch], repo_root, git_timeout)
+        except subprocess.TimeoutExpired:
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/self_update.py:348]: Addressed in 46c850d. Added an `_is_safe_arg` guard: `remote` and `branch` are validated before any git call and rejected when empty or starting with `-`, returning a new `SKIPPED_BAD_CONFIG` outcome. This prevents a misconfigured `AGENTS_AUTO_UPDATE_REMOTE` / `AGENTS_AUTO_UPDATE_BRANCH` from being parsed as a git option.

File: tests/test_self_update.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+import os
+import subprocess
+from pathlib import Path
+from types import SimpleNamespace
+
+import pytest
+
+from src import self_update
+from src.self_update import UpdateStatus, check_and_apply_update
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_self_update.py:16]: Addressed in 46c850d. Added a module-level `pytestmark = pytest.mark.skipif(shutil.which("git") is None, ...)` so the suite skips cleanly instead of erroring when the git CLI is not on PATH.

File: src/self_update.py [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+        new_sha = new.stdout.strip() if new.returncode == 0 else ""
+        logger.info("Auto-update: fast-forwarded %s -> %s on %s.", old_sha[:9], new_sha[:9], branch)
+
+        _warn_if_deps_changed(repo_root, old_sha, new_sha, git_timeout)
+
+        # Step 7 — rebuild the vector stores with the new code; roll back on failure.
+        if not reindex(repo_root):
+            logger.error("Auto-update: reindex failed; rolling back to %s.", old_sha[:9])
+            rollback = _run_git(["reset", "--hard", old_sha], repo_root, git_timeout)
+            if rollback.returncode != 0:
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/self_update.py:411]: Not applied, after verifying the behavior locally. `git reset --hard <old_sha>` reverts tracked files, and files added by the fast-forwarded commits are tracked, so the reset removes them: after the rollback the added files and directories are gone and `git status --porcelain` is empty. A pre-existing untracked file that would collide with an incoming path makes `git merge --ff-only` fail (MERGE_FAILED) before any reindex or rollback, so the rollback cannot leave added files behind. The added-file path is covered by `test_reindex_failure_rollback_removes_added_file`.

File: tests/test_self_update.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    _commit(repos.upstream, "file.txt", "v2\n", "update")
+    old = _head(repos.local)
+
+    status = check_and_apply_update(
+        str(repos.local), "origin", "main", reindex_fn=lambda rr: False
+    )
+
+    assert status == UpdateStatus.REINDEX_FAILED
+    assert _head(repos.local) == old  # rolled back to the pre-merge commit
+
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_self_update.py:109]: Addressed in 46c850d. Added `test_reindex_failure_rollback_removes_added_file`: the upstream update adds a new path, the reindex fails, and the test asserts the file is removed and the working tree is clean after rollback.

File: src/self_update.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+        if status.returncode != 0:
+            logger.warning("Auto-update: git status failed; skipping. %s", status.stderr.strip())
+            return UpdateStatus.SKIPPED_DIRTY
+        if status.stdout.strip():
+            logger.info("Auto-update: working tree has uncommitted changes; skipping.")
+            return UpdateStatus.SKIPPED_DIRTY
+
+        head = _run_git(["rev-parse", "HEAD"], repo_root, git_timeout)
+        old_sha = head.stdout.strip() if head.returncode == 0 else ""
+
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/self_update.py:343]: Addressed in 46c850d, both points. The updater now returns early before fetch/merge when HEAD cannot be resolved, so `git reset --hard` is never invoked with an empty SHA. The `reindex()` call is wrapped so an exception is treated as a failure and the rollback still runs. Covered by `test_reindex_exception_triggers_rollback`.

File: src/self_update.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+            return UpdateStatus.FETCH_FAILED
+        if behind == 0:
+            logger.info("Auto-update: already up to date with %s.", remote_ref)
+            return UpdateStatus.UP_TO_DATE
+        if ahead > 0:
+            logger.info(
+                "Auto-update: local branch diverged from %s (ahead %d, behind %d); skipping.",
+                remote_ref, ahead, behind,
+            )
+            return UpdateStatus.SKIPPED_DIVERGED
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply src/self_update.py]: Fixed in 5d3735f. The ahead check now runs before the up-to-date check: when the local branch is ahead with nothing to pull (ahead > 0, behind == 0), the updater returns a new `SKIPPED_AHEAD` outcome instead of `UP_TO_DATE`. The previous behavior was already a no-op (no fetch-merge, no mutation), so this is a status and logging correctness change; local commits are never fast-forwarded over.

File: tests/test_self_update.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+import os
+import subprocess
+from pathlib import Path
+from types import SimpleNamespace
+
+import pytest
+
+from src import self_update
+from src.self_update import UpdateStatus, check_and_apply_update
+
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_self_update.py:17]: Already handled in 46c850d (this review re-ran against the earlier commit 88f8de2). The test module now sets `pytestmark = pytest.mark.skipif(shutil.which("git") is None, ...)`, so the suite skips instead of erroring when git is not on PATH.

File: tests/test_self_update.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    _commit(repos.upstream, "file.txt", "remote-change\n", "remote")
+    _commit(repos.local, "other.txt", "local-change\n", "local")
+
+    status = check_and_apply_update(str(repos.local), "origin", "main", reindex_fn=recorder)
+
+    assert status == UpdateStatus.SKIPPED_DIVERGED
+    assert recorder.calls == []
+    # local commit is preserved
+    assert _head(repos.local) != _head(repos.upstream)
+
```
[the remark this answers is not in the document]
Comment (WonderMr) [reply tests/test_self_update.py:176]: Fixed in 5d3735f. Added `test_skip_when_ahead_only`: the local branch has a commit the remote lacks (ahead > 0, behind == 0); the test asserts the updater returns `SKIPPED_AHEAD`, does not reindex, and leaves HEAD untouched.

**Comment by WonderMr** [comment]
Re the CodeRabbit outside-diff finding on `src/self_update.py` (throttling after a pre-fetch git timeout): fixed in 5d3735f. `check_and_apply_update` now tracks a `network_reached` flag set right before `git fetch`. The outer `except subprocess.TimeoutExpired` returns `FETCH_FAILED` only when the network step was reached; a pre-fetch local-op timeout (`rev-parse` / `status`) returns `NO_GIT`, which is a pre-network status and does not write the throttle stamp, so a purely local timeout no longer suppresses the next real update attempt.

**Comment by WonderMr** [comment]
Re the CodeRabbit outside-diff findings on `src/self_update.py` and the tests ("Outside diff range comments (3)"), addressed in 6bbfa42:

**O1 (post-merge rollback, Major): fixed.** The function now tracks a `merged` flag set right after the fast-forward, and the outer handler catches both `subprocess.TimeoutExpired` and `FileNotFoundError`. If an exception occurs after the merge, it runs a best-effort `git reset --hard <old_sha>` before returning, so a post-merge timeout or a vanished git binary can no longer strand the install on new code without rebuilt indexes. Covered by `test_exception_after_merge_rolls_back`.

**O3 (timeout/throttle coverage, Nitpick): added.** `test_timeout_before_fetch_is_pre_network` (a pre-fetch timeout returns NO_GIT and does not write the throttle stamp), `test_timeout_after_fetch_is_network_failure` (a post-fetch timeout returns FETCH_FAILED), and `test_network_reached_failure_writes_throttle_stamp`.

**O2 (persist every terminal outcome, Minor): not applied.** `.last_update.json` records the last applied update or rollback by design, not every check, so on the next startup it correctly reports the last real update even when the current run is up-to-date or skipped. Writing state on every early-return path would also duplicate `_write_state` across roughly fourteen branches. The startup log line was reworded to "last applied update" to make that scope unambiguous.
